File: 240120/최대 증가 부분 수열/longest-increasing-subsequence.cpp
#include <iostream>

using namespace std;

int arr[1000] = {0};

int main(){
    int n;
    cin >> n;

    int dp[1000] = {0};
    dp[0] = 1;

    for(int i=0; i<n; i++){
        cin >> arr[i];
    }

    for(int i = 1; i<n; i++){
        for(int j=i-1; j>=0; j--){
            if((arr[j] < arr[i]) && (dp[i] < dp[j]+1)) {
                dp[i] = dp[j]+1;
            }
        }
    }

    int max = 1;
    for(int i = 1; i<n; i++){
        if (max < dp[i]) max = dp[i];
    }
    
    cout << max << '\n';
}

File: 251011/최대 증가 부분 수열/longest-increasing-subsequence.py
# n = int(input())
# arr = list(map(int, input().split()))
# dp = [-1 for _ in range(n)]

# def recur(idx):
#     if dp[idx] != -1 : return dp[idx]

#     cnt = 1

#     for i in range(idx+1,n):
#         if  arr[idx] < arr[i] :
#             cnt = max(cnt, recur(i)+1)
    
#     dp[idx] = cnt
#     return dp[idx]

# ans = 1
# for i in range(n):
#     ans = max(recur(i), ans)

# print(ans)

N = int(input())

nums = list(map(int,input().split()))

dp = [-1 for _ in range(N)]

dp[0] = 1

for i in range(1,N):
    max_dp = -1
    for j in range(i):
        if nums[j] < nums[i] : 
            max_dp = dp[j] if max_dp < dp[j] else max_dp
    
    dp[i] = max_dp + 1 if max_dp > 0 else 1

print(max(dp))